service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public void getTagsForCertificates(GiftCertificate giftCertificates) {
+        List<Tag> tagsForCertificate = tagDao.getTagsForCertificate(giftCertificates.getId());
+        giftCertificates.setTags(tagsForCertificate);
+    }
+
+    @Override
+    public GiftCertificate findById(long id) {
+        GiftCertificate giftCertificate = giftCertificateDao.getById(id);
+        List<Tag> tagsForCertificate = tagDao.getTagsForCertificate(id);
+        giftCertificate.setTags(tagsForCertificate);
+        return giftCertificate;
+    }
+
+    @Override
+    @Transactional
+    public GiftCertificate save(GiftCertificate giftCertificate) {
+        LocalDateTime localDateTime = LocalDateTime.now(zoneId);
+        giftCertificate.setCreateDate(localDateTime);
+        giftCertificate.setLastUpdateDate(localDateTime);
+
+        GiftCertificate savedGiftCertificate = giftCertificateDao.create(giftCertificate);
+        long id = savedGiftCertificate.getId();
+
+        saveNewTags(savedGiftCertificate);
+        assignTagsToCertificate(id, giftCertificate.getTags());
+        List<Tag> tagsForCertificate = tagDao.getTagsForCertificate(id);
+        savedGiftCertificate.setTags(tagsForCertificate);
+        return savedGiftCertificate;
+    }
+
+    @Override
+    @Transactional
+    public void delete(long id) {
+        giftCertificateDao.remove(id);
+    }
+
+    @Override
+    @Transactional
+    public void update(GiftCertificate giftCertificate) {
+        GiftCertificateUpdateValidator.validate(giftCertificate);
+        long certificateId = giftCertificate.getId();
+        if (giftCertificate.getTags() != null) {
+            saveNewTags(giftCertificate);
+            tagDao.detachTagsFromCertificate(certificateId);
+            assignTagsToCertificate(certificateId, giftCertificate.getTags());
+            giftCertificate.setTags(tagDao.getTagsForCertificate(certificateId));
+        }
+        giftCertificate.setLastUpdateDate(LocalDateTime.now(zoneId));
+        giftCertificateDao.update(giftCertificate);
+    }
+
+    private void assignTagsToCertificate(Long certificateId, List<Tag> tags) {
+        if (tags == null || tags.isEmpty()) return;
+        List<Tag> tagsToAssign = getIdsForTags(tags);
+        for (Tag tag : tagsToAssign) {
+            tagDao.assignTagToCertificate(certificateId, tag.getId());
+        }
+    }
+
+    private List<Tag> getIdsForTags(List<Tag> tags) {
+        List<Tag> tagsWithIds = new ArrayList<>(tags.size());
+        for (Tag newTag : tags) {
+            Optional<Tag> tagWithId = tagDao.getByName(newTag.getName());
+            tagWithId.ifPresent(tagsWithIds::add);
+        }
+        return tagsWithIds;
+    }
+
+    private void saveNewTags(GiftCertificate giftCertificate) {
+        List<Tag> newTags = giftCertificate.getTags();
+        if (newTags == null || newTags.isEmpty())
+            return;
+
+        List<Tag> allTags = tagDao.getAll();
+
+        for (Tag newTag : newTags) {
```

Comment:
Why dont you use something like

for (var tag : newTags){
tagService.findByName(tag.getName())
.ifPresentOrElse(...)
}

would be much readable